game-app/game-core/src/main/java/games/strategy/triplea/delegate/Matches.java

```diff
@@ -1146,39 +1138,21 @@ public static Predicate<Unit> unitHasEnoughMovementForRoute(final Route route) {
     return unit -> {
       BigDecimal left = unit.getMovementLeft();
       final UnitAttachment ua = UnitAttachment.get(unit.getType());
-      final GamePlayer player = unit.getOwner();
       if (ua.getIsAir()) {
-        final TerritoryAttachment taStart = TerritoryAttachment.get(route.getStart());
-        final TerritoryAttachment taEnd = TerritoryAttachment.get(route.getEnd());
-        if (taStart != null && taStart.getAirBase()) {
+        if (TerritoryAttachment.hasAirBase(route.getStart())) {
           left = left.add(BigDecimal.ONE);
         }
-        if (taEnd != null && taEnd.getAirBase()) {
+        if (TerritoryAttachment.hasAirBase(route.getEnd())) {
           left = left.add(BigDecimal.ONE);
         }
       }
-      final GameStep stepName = unit.getData().getSequence().getStep();
-      if (ua.getIsSea() && stepName.getDisplayName().equals("Non Combat Move")) {
+      if (ua.getIsSea() && unit.getData().getSequence().getStep().isNonCombat()) {
         // If a zone adjacent to the starting and ending sea zones are allied naval bases, increase
         // the range.
         // TODO Still need to be able to handle stops on the way
-        // (history to get route.getStart()
-        for (final Territory terrNext : unit.getData().getMap().getNeighbors(route.getStart(), 1)) {
-          final TerritoryAttachment taNeighbor = TerritoryAttachment.get(terrNext);
-          if (taNeighbor != null
-              && taNeighbor.getNavalBase()
-              && unit.getData().getRelationshipTracker().isAllied(terrNext.getOwner(), player)) {
-            for (final Territory terrEnd :
-                unit.getData().getMap().getNeighbors(route.getEnd(), 1)) {
-              final TerritoryAttachment taEndNeighbor = TerritoryAttachment.get(terrEnd);
-              if (taEndNeighbor != null
-                  && taEndNeighbor.getNavalBase()
-                  && unit.getData().getRelationshipTracker().isAllied(terrEnd.getOwner(), player)) {
-                left = left.add(BigDecimal.ONE);
-                break;
-              }
-            }
-          }
+        if (hasNeighboringAlliedNavalBase(route.getStart(), unit.getOwner())
+            && hasNeighboringAlliedNavalBase(route.getEnd(), unit.getOwner())) {
+          left = left.add(BigDecimal.ONE);
```

Comment:
Doing a self-review, I wasn't sure if this was quite right. The previous logic looped through neighbours of the start territory to find naval bases and for each, looped through neighbors of an end territory to find naval bases.

Theoretically, it could have increased movement by more than one (under very specific circumstances), but this was actually a bug!

According to https://axisandallies.fandom.com/wiki/Naval_Base, the logic should be:
>  Adds 1 to non-combat sea movement, if the ships starts and ends adjacent to friendly naval bases

Which matches what the new implementation does.


<br class="Apple-interchange-newline">

Reply:
However, the above link also says that the logic should be different for "AAG40" rule set. I don't see this here so I'm wondering if that's a missing implementation in TripleA?

Reply:
I guess TripleA doesn't support "Axis & Allies: WWI 1914" - either the map or the rule set?

Reply:
So just to be clear, the logic change is as follows:

Previously, if a sea unit was at a territory adjacent to multiple naval bases and had a route to a territory next to another naval base, then it's movement would be increased by the number of naval bases it was adjacent to! That doesn't seem to match the rules.

Curiously, looking at the actual rules from 1940 global 2nd edition, it doesn't seem to talk about the route *ending* next to a naval base:

https://www.axisandallies.org/wp-content/uploads/2011/12/Axis-Allies-Europe-1940-Second-Edition.pdf
> 
> Increased Sea Unit Range: All sea units beginning their
> movement from a sea zone serviced by an operative
> friendly naval base gain 1 additional point of movement
> range. Sea units in this position can move 3 spaces
> instead of 2.

So, why do we have different logic here?

Reply:
So, in-game it doesn't seem to work this way - it seems to match the PDF rules (no need for naval base at destination and no multiple bonuses). Perhaps the "source of truth" logic is elsewhere and not here? I can dig more.

If so, it also means there's a bug in this logic.

Reply:
So, searching around, I found https://github.com/triplea-game/triplea/issues/9584 which mentions the other version of Naval bases which matches the wiki. It quotes a manual that I found here:
https://kipdf.com/gameplay-manual-players-2-to-3-ages-12-and-up-what-s-different-about-axis-allies_5ae431a97f8b9ab3818b45b5.html

> During the Non-combat Movement phase, naval units moving from a sea zone containing a friendly naval base to another sea zone containing a friendly naval base may move three spaces instead of two. Either, or both, of these naval bases may have been captured on this turn. 

But https://github.com/triplea-game/triplea/issues/9584 seems to suggest the current engine doesn't support this. So this code is quite curious.

Reply:
OK, so after digging a bit more, here are my findings:
  - This code actually does work (and has the bug I mentioned which this is fixing), when there is actually a "Naval base" present.
  - However, most maps don't use this "Naval base" feature - instead, Naval bases on maps like 1940 Global instead just boost units movement using a different function.
  - On "Global War 2", this feature is used: https://github.com/search?q=org%3Atriplea-maps+navalBase&type=code
  - On that map, we can see the bug in action for the British ships in SZ 8, which has two naval bases neighbouring. We can see that the ships there can move 4 spaces to SZ 5, instead of 3 - and with this fix, it becomes 3 as it should be.
  - Having said that, there is no test of this code at all - just removing this logic doesn't fail anything.

So per the above, I'm going to extract this specific part of the PR into a separate one, so this one is kept as "no functional changes" - and the other one can focus on this improvement along with some tests.